src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RSACryptoServiceProvider.cs

```diff
@@ -64,29 +65,99 @@ private RSACryptoServiceProvider(int keySize, CspParameters parameters, bool use
+                            Debug.Assert(!hProv.IsInvalid);
+                            Debug.Assert(!hProv.IsClosed);
+
+                            _safeProvHandle = hProv;
+                        }
+                    }
+
+                    return _safeProvHandle;
+                }
+
+                return _safeProvHandle;
+            }
+            set
             {
-                lock (this)
+                lock (_parameters)
                 {
-                    if (_safeKeyHandle == null)
+                    if (ReferenceEquals(value, _safeProvHandle))
                     {
-                        // We only attempt to generate a random key on desktop runtimes because the CoreCLR
-                        // RSA surface area is limited to simply verifying signatures.  Since generating a
-                        // random key to verify signatures will always lead to failure (unless we happend to
-                        // win the lottery and randomly generate the signing key ...), there is no need
-                        // to add this functionality to CoreCLR at this point.
-                        CapiHelper.GetKeyPairHelper(CapiHelper.CspAlgorithmType.Rsa, _parameters,
-                                            _randomKeyContainer, _keySize, ref _safeProvHandle, ref _safeKeyHandle);
+                        return;
                     }
+
+                    if (_safeProvHandle != null)
+                    {
+                        _safeKeyHandle?.Dispose();
+                        _safeKeyHandle = null;
+
+                        _safeProvHandle?.Dispose();
+                    }
+
+                    _safeProvHandle = value;
+                }
+            }
+        }
+
+        private SafeKeyHandle SafeKeyHandle
+        {
+            get
+            {
+                if (_safeKeyHandle == null)
+                {
+                    lock (_parameters)
+                    {
+                        if (_safeKeyHandle == null)
+                        {
+                            SafeKeyHandle hKey = CapiHelper.GetKeyPairHelper(
+                                CapiHelper.CspAlgorithmType.Rsa,
+                                _parameters,
+                                _keySize,
+                                SafeProvHandle);
+
+                            Debug.Assert(hKey != null);
+                            Debug.Assert(!hKey.IsInvalid);
+                            Debug.Assert(!hKey.IsClosed);
+
+                            _safeKeyHandle = hKey;
+                        }
+                    }
+                }
+
+                return _safeKeyHandle;
+            }
+
+            set
+            {
+                lock (_parameters)
+                {
+                    if (ReferenceEquals(value, _safeKeyHandle))
+                    {
+                        return;
+                    }
+
+                    _safeKeyHandle?.Dispose();
```

Comment:
Can this race with the getter and end up with the getter returning a disposed key handle? I think we typically save the value into a local and then overwrite the field to prevent that kind of thing (though I'm still not quite sure that's right here)